src/main/java/uk/gov/hmcts/reform/blobrouter/services/storage/BulkScanSasTokenCache.java

```diff
@@ -0,0 +1,93 @@
+import java.util.concurrent.TimeUnit;
+
+import static com.azure.storage.common.implementation.Constants.UrlConstants.SAS_EXPIRY_TIME;
+import static com.azure.storage.common.implementation.StorageImplUtils.parseQueryString;
+import static java.time.temporal.ChronoField.INSTANT_SECONDS;
+import static uk.gov.hmcts.reform.blobrouter.exceptions.InvalidSasTokenException.EXPIRY_NOT_FOUND;
+
+@Service
+public class BulkScanSasTokenCache {
+
+    private final BulkScanProcessorClient bulkScanSasTokenClient;
+    private final long refreshSasBeforeExpiry;
+
+    //key= container name, value = sastoken
+    private static Cache<String, String> tokenCache;
+
+    public BulkScanSasTokenCache(
+        BulkScanProcessorClient bulkScanSasTokenClient,
+        @Value("${bulk-scan-cache.refresh-before-expire-in-sec}") long refreshSasBeforeExpiry
+    ) {
+        this.bulkScanSasTokenClient = bulkScanSasTokenClient;
+        this.refreshSasBeforeExpiry = refreshSasBeforeExpiry;
+        tokenCache = Caffeine.newBuilder()
+            .expireAfter(new BulkScanSasTokenCacheExpiry())
+            .build();
+    }
+
+    public String getSasToken(String containerName) {
+        return tokenCache.get(containerName, conName -> this.createSasToken(conName));
+    }
+
+    private String createSasToken(String containerName) {
+        return bulkScanSasTokenClient.getSasToken(containerName).sasToken;
+    }
+
+    private class BulkScanSasTokenCacheExpiry implements Expiry<String, String> {
+
+        @Override
+        public long expireAfterCreate(
+            @NonNull String containerName,
+            @NonNull String sasToken, long currentTime) {
+            Map<String, String> map = parseQueryString(sasToken);
+            return calculateTimeToExpire(
+                Constants.ISO_8601_UTC_DATE_FORMATTER.parse(
+                    map.computeIfAbsent(
+                        SAS_EXPIRY_TIME, key -> {
+                            throw new InvalidSasTokenException(EXPIRY_NOT_FOUND);
+                        }
+                    )
+                )
+            );
+        }
+
+        @Override
+        public long expireAfterUpdate(
+            @NonNull String containerName, @NonNull String sasToken, long currentTime,
+            @NonNegative long currentDuration) {
+            return expireAfterCreate(containerName, sasToken, currentTime);
+        }
+
+        @Override
+        public long expireAfterRead(
+            @NonNull String containerName, @NonNull String sasToken, long currentTime,
+            @NonNegative long currentDuration) {
+            return currentDuration;
+        }
+
+        private long calculateTimeToExpire(TemporalAccessor expiry) {
+            return
+                TimeUnit.NANOSECONDS.convert(
+                    expiry.getLong(INSTANT_SECONDS)
+                        - (OffsetDateTime.now(ZoneOffset.UTC).getLong(INSTANT_SECONDS) + refreshSasBeforeExpiry),
+                    TimeUnit.SECONDS);
```

Comment:
Suggestion: consider defining some method-local variables on the way, to better describe what's going on here.

Reply:
worth doing this as it is not that readable

Reply:
added method javadoc